### DISCUSSION TITLE
mqtt sink server param no longer accepts path

### DISCUSSION THREAD

**Discussion by wtflanders** [lf-edge/ekuiper#2827] [Q&A]
Hi, it's been a while since I updated my eKuiper version, I've been running on v1.4.2.  The "server" param of an mqtt sink rule with 1.4.2 could have a /<path> on the endpoint:  
  "server": "wss://hostname.x.com:443/mqtt"

Testing v1.13.3 I get a "server is missing" error when I add the `/mqtt` path portion to the endpoint.... and my connecting broker requires it.  
Is there a way around this?
When did it change?
Thx.

Here is my rule:
{ "ID": "postAlive",   "SQL": "SELECT ts FROM alive_stream",   "Actions": [ {
                        "mqtt": {
                                "server": "wss://host.x.com:443/mqtt",
                                "clientId": "alive-host.micron.com",
                                "password": "abcdefg",
                                "sendSingle": true,
                                "topic": "host/alive",
                                "username": "newusr"
                        }
                }
        ]
}

**Comment by wtflanders** [comment]
Here is the rule status:

  ```
{

          "id": "postAlive",
          "name": "postAlive",
          "status": "Stopped: missing server property."
  }
```

**Comment by Yisaer** [comment]
Hi @wtflanders ,

I can't reproduce the same error in the 1.13.3 with your rule. Could you check with the rule spec by rest api `GET /rules/{ruleID}` which reported the error `missing server property`. This error indicates that the server is not defined, in another word, the server is "".

**Comment by wtflanders** [comment]
Hi @Yisaer,
Hi,
Thanks for the reply.

Here is the actual return text when creating rule using the kuiper CLI -

This one works, from lfedge/ekuiper:1.4.2-slim-python docker image:
```
        root@edge40-kuiper-0:/kuiper/bin# ./kuiper create rule  postAlive2 '{ "ID": "postAlive2",
                "SQL": "SELECT ts FROM alive_stream",
                "Actions": [
                        {
                                "mqtt": {
                                        "clientId": "abc1",
                                        "dataTemplate": "{{ .ts}}",
                                        "password": "???",
                                        "sendSingle": true,
                                        "server": "wss://mirror.company.com:443/mqtt",
                                        "topic": "data/host1.company.com/edge/alive",
                                        "username": "cousr"
                                }
                        }
                ]
        }'
        Connecting to 127.0.0.1:20498... 
        Rule postAlive2 was created successfully, please use 'bin/kuiper getstatus rule postAlive2' command to get rule status.
```

This one fails, from lfedge/ekuiper:1.13.3-slim-python docker image:

```
        root@edge40-kuiper-0:/kuiper/bin# ./kuiper create rule  postAlive2 '{ "ID": "postAlive2",
                "SQL": "SELECT ts FROM alive_stream",
                "Actions": [
                        {
                                "mqtt": {
                                        "clientId": "abc2",
                                        "dataTemplate": "{{ .ts}}",
                                        "password": "???",
                                        "sendSingle": true,
                                        "server": "wss://mirror.company.com:443/mqtt",
                                        "topic": "data/host2.company.com/edge/alive",
                                        "username": "cousr"
                                }
                        }
                ]
        }'
        Connecting to 127.0.0.1:20498... 
        Create rule postAlive2 error : missing server property.
```

let me know if you need more info,
Bill

**Comment by gexincoding** [comment]
@wtflanders 
Thank you for reporting this issue. I've thoroughly reviewed the debug logs and the associated code, and everything seems to be functioning as expected on our end. It's possible that the issue might be related to the CLI tool you are using.

Could you please try accessing the functionality through the HTTP API instead and see if the issue persists?

**Comment by wtflanders** [comment]
Stream cmd works:
```
curl -X POST http://localhost:9081/streams 'Content-Type: application/json' -d '{ "sql":"create stream alive_stream ( ) WITH (app/alive\", FORMAT= \"json\", KEY = \"id\")" }'
Stream alive_stream2 is created.

curl http://localhost:9081/streams 
["alive_stream"]
```

But rule cmd fails with same error message:
```
curl -X POST http://localhost:9081/rules 'Content-Type: application/json' -d '{ "ID": "postAlive2",
>                 "SQL": "SELECT ts FROM alive_stream",
>                 "Actions": [
>                         {
>                                 "mqtt": {
>                                         "clientId": "abc2",
>                                         "dataTemplate": "{{ .ts}}",
>                                         "password": "???",
>                                         "sendSingle": true,
>                                         "server": "wss://mirror.company.com:443/mqtt",
>                                         "topic": "data/host2.company.com/edge/alive",
>                                         "username": "cousr"
>                                 }
>                         }
>                 ]
>         }'
{"error":2201,"message":"missing server property"}
```

Thanks!

**Reply by wtflanders** [reply]
version:
```
curl http://localhost:9081/
{"version":"1.13.3","os":"linux","arch":"amd64","upTimeSeconds":3091,"cpuUsage":"0.02%","memoryUsed":"69042176","memoryTotal":"9223372036854771712"}[wtflanders@sigdev08 ~]$ http_proxy="" && curl http://localhost:9081/ | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   148  100   148    0     0  74000      0 --:--:-- --:--:-- --:--:-- 74000
{
  "version": "1.13.3",
  "os": "linux",
  "arch": "amd64",
  "upTimeSeconds": 3106,
  "cpuUsage": "0.02%",
  "memoryUsed": "69042176",
  "memoryTotal": "9223372036854771712"
}
```

**Reply by wtflanders** [reply]
More info-  v1.8.1 successfully creates the rule but then the rule FAILS with a similar error message:
```
[
  {
    "id": "postAlive2",
    "name": "postAlive2",
    "status": "Stopped: missing server property."
  }
]
```

**Reply by wtflanders** [reply]
I narrowed it down, v1.4.3 works but v1.5.1 does not, same issue as 1.8.1.    These are all the slim-python images if that matters.

**Comment by gexincoding** [comment]
Please try modifying the format of the JSON parameter in your curl command or command line. Try to keep the JSON structure intact without line breaks.

![image](https://github.com/lf-edge/ekuiper/assets/98001140/d904c436-4d39-434c-93e4-d68ba4b317a8)

{
"version": "1.13.1",
"os": "linux",
"arch": "arm64",
"upTimeSeconds": 1126,
"cpuUsage": "0.06%",
"memoryUsed": "102047744",
"memoryTotal": "8326725632"
}

**Comment by wtflanders** [comment]
I simplified it, still no luck:

curl -X POST http://localhost:9081/rules 'Content-Type: application/json' -d '{"ID":"postAlive2","SQL":"SELECT ts FROM alive_stream","Actions":[{"mqtt":{"server":"ws://dev00.company.com:8093/mqtt","topic":"ktest/dev/edge/alive"}}]}'
**{"error":2201,"message":"missing server property"}**

This works with v1.4.2.
Thanks.

**Reply by gexincoding** [reply]
Please use the get request to get version information.

Mine is：
{
"version": "1.13.1",
"os": "linux",
"arch": "arm64",
"upTimeSeconds": 78,
"cpuUsage": "0.43%",
"memoryUsed": "104681472",
"memoryTotal": "8326725632"
}


my curl is
curl -X POST http://localhost:9081/rules 'Content-Type: application/json' -d '{"ID":"postAlive3","SQL":"SELECT * FROM demo","Actions":[{"mqtt":{"server":"ws://dev00.company.com:8093/mqtt","topic":"ktest/dev/edge/alive"}}]}'

it works：

![image](https://github.com/lf-edge/ekuiper/assets/98001140/aa03f120-b36b-4d85-a6fd-caeca9b4c5c1)


Please try to check the version

**Comment by wtflanders** [comment]
curl http://localhost:9081 | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   147  100   147    0     0   143k      0 --:--:-- --:--:-- --:--:--  143k
{
  "version": "1.13.1",
  "os": "linux",
  "arch": "amd64",
  "upTimeSeconds": 533,
  "cpuUsage": "0.05%",
  "memoryUsed": "68759552",
  "memoryTotal": "9223372036854771712"
}

**Comment by wtflanders** [comment]
@gexincoding  I have eKuiper deployed via a helm chart in a kubernetes cluster.   I scaled the replicas in that cluster down to 0 then ran 1.13.1 outside of the cluster via docker.  It now works.  

I went back and dug through the v1.5 rel notes, it turns it out it was the **servers** param!  Which... seemed to have nothing to do with the rules but when I changed the eKuiper deployment to use **server** instead of **servers** it now works in my cluster!!!

Thank you for all your help.  I had seen this change when I first posted but did not think it was related.
